ciao-cli/template.go

```diff
@@ -213,3 +215,120 @@ func createTemplate(name, tmplSrc string) *template.Template {
 
 	return t
 }
+
+func exportedFields(typ reflect.Type) bool {
+	var i int
+	for i = 0; i < typ.NumField(); i++ {
+		if typ.Field(i).PkgPath == "" {
+			break
+		}
+	}
+
+	return i < typ.NumField()
+}
+
+func ignoreKind(kind reflect.Kind) bool {
+	return (kind == reflect.Chan) || (kind == reflect.Invalid)
+}
+
+func generateStruct(buf io.Writer, dummy reflect.Type) {
+	for i := 0; i < dummy.NumField(); i++ {
+		field := dummy.Field(i)
+		if field.PkgPath != "" || ignoreKind(field.Type.Kind()) {
+			continue
+		}
+		fmt.Fprintf(buf, "%s ", field.Name)
+		tag := ""
+		if field.Tag != "" {
+			tag = fmt.Sprintf("`%s`", field.Tag)
+		}
+		generateUsage(buf, field.Type, tag)
+	}
+}
+
+// This function is not currently used.  The idea was to annotate
+// types with their user defined typenames, if such names existed.
+// This would allow users to know whether they could invoke
+// methods on those types.  Unfortunately, I couldn't figure out
+// a nice way of displaying this information, so it's disabled for
+// now.
+func getFullTypeName(typ reflect.Type) string {
+	if typ.Name() == "" {
+		return ""
+	}
+	pkg := typ.PkgPath()
+	if pkg != "" {
+		li := strings.LastIndex(pkg, "/")
+		if li != -1 {
+			pkg = pkg[li+1:]
+		}
+		pkg = fmt.Sprintf("%s.", pkg)
+	}
+	return fmt.Sprintf("%s%s", pkg, typ.Name())
+}
+
+func generateUsage(buf io.Writer, dummy reflect.Type, tag string) {
```

Comment:
The name dummy was chosen to reflect the fact that the value between passed to the function exists only to convey type information.  It's value is unimportant.  The only other example of a function like this i can find in the standard library is reflect.TypeOf, whose parameter is

i interface{}

I'll change the code to match.